syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -14,145 +15,175 @@
-                blocksByHash.clear()
-            }
-        }
-        synchronized(lock) {
-            hashList.addAll(fileBlocks.blocks.map { it.hash })
-            missingHashes.addAll(hashList)
-            for (block in fileBlocks.blocks) {
-                if (missingHashes.contains(block.hash)) {
-                    val requestId = Math.abs(Random().nextInt())
-                    requestIds.add(requestId)
-                    connectionHandler.sendMessage(Request.newBuilder()
-                            .setId(requestId)
-                            .setFolder(fileBlocks.folder)
-                            .setName(fileBlocks.path)
-                            .setOffset(block.offset)
-                            .setSize(block.size)
-                            .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
-                            .build())
-                    logger.debug("sent request for block, hash = {}", block.hash)
+                fileBlocks.blocks.distinctBy { it.hash }.forEach { block ->
+                    pipe.send(block)
                 }
-            }
-            return fileDownloadObserver
-        }
-    }
 
-    fun onResponseMessageReceived(response: BlockExchangeProtos.Response) {
-        synchronized(lock) {
-            if (!requestIds.contains(response.id)) {
-                return
+                pipe.close()
             }
-            NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR, {"received error response, code = ${response.code}"})
-            val data = response.data.toByteArray()
-            val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
-            if (missingHashes.remove(hash)) {
-                blocksByHash.put(hash, data)
-                logger.debug("aquired block, hash = {}", hash)
-                lock.notify()
-            } else {
-                logger.warn("received not-needed block, hash = {}", hash)
+
+            // the sequence is evaluated lazy -> only one block per time is loaded
+            val fileBlocksIterator = fileBlocks.blocks
+                    .asSequence()
+                    .map { tempRepository.popTempData(blockTempIdByHash[it.hash]!!) }
+                    .map { ByteArrayInputStream(it) }
+                    .iterator()
+
+            return object : SequenceInputStream(object : Enumeration<InputStream> {
+                override fun hasMoreElements() = fileBlocksIterator.hasNext()
+                override fun nextElement() = fileBlocksIterator.next()
+            }) {
+                override fun close() {
+                    super.close()
+
+                    // delete all temp blocks now
+                    // they are deleted after reading, but the consumer could stop before reading the whole stream
+                    tempRepository.deleteTempData(blockTempIdByHash.values.toList())
+                }
             }
+        } catch (ex: Exception) {
```

Comment:
You shouldn't catch general exceptions, that might hide things like a NullPointerException.

Reply:
Ah I see. I also missed that you are re-throwing the exception, so this is fine.